### PR TITLE
Rm transfer tokentype

### LINKED ISSUE #206: Trade: Commit Fee Receiver's tokenType or Add one more witness in the fraud proof
## Summary

We either commit token_type of the fee receiver, or send one more witness in the fraud proof.

## Helpers

```python
def require_coordinator(predicate: bool):
    # Return invalid batch in Solidity
    if not predicate:
        return "Slash coordinator"

def require_dispute(predicate: bool):
    # `require` in Solidity
    if not predicate:
        return "Bad Dispute"
```

## Case 1: commit tokenType

Overhead: 1 uint256 token_type to commitment

```python
def process_commit(commit, dispute):
    token_type = commit.token_type
    for tx in commit.txs:
        require_coordinator(tx.token_type == token_type)
        state_root, fee_sum = process_tx(tx)
    
    require_dispute(
        merkle_check(
            state_root,
            commit.fee_index,
            dispute.fee_account,
            dispute.fee_witness
        )
    )
    require_coordinator(dispute.fee_account.token_type == token_type)
    state_root = apply_fee(dispute.fee_account, fee_sum)
```

### Case2.1: no validate fee_account at the start

Problem: Bad dispute slash good coordinator by sending a bad fee_account.

```python
def process_commit(commit, dispute):
    token_type = dispute.fee_account.token_type

    for tx in commit.txs:
        require_coordinator(tx.token_type == token_type)
        state_root, fee_sum = process_tx(tx)
    
    require_dispute(
        merkle_check(
            state_root,
            commit.fee_index,
            dispute.fee_account,
            dispute.fee_witness
        )
    )
    state_root = apply_fee(commit.fee_account, fee_sum)
```

### Case2.2: Validate fee_account at the start

Problem: fee_account can send free token in txs.

```python
def process_commit(commit, dispute):
    require_dispute(
        merkle_check(
            state_root,
            commit.fee_index,
            dispute.fee_account,
            dispute.fee_witness
        )
    )
    token_type = commit.fee_account.token_type

    for tx in commit.txs:
        require_coordinator(tx.token_type == token_type)
        state_root, fee_sum = process_tx(tx)

    state_root = apply_fee(dispute.fee_account, fee_sum)
```

### Case2.3: Validate fee_account at the start and at the end

Overhead: need 1 more witness in fraud proof

```python
def processCommit(commit):
    require_dispute(
        merkle_check(
            state_root,
            commit.fee_index,
            dispute.fee_account_pre,
            dispute.fee_witness_pre
        )
    )
    token_type = commit.fee_account.token_type

    for tx in commit.txs:
        require_coordinator(tx.token_type == token_type)
        state_root, fee_sum = process_tx(tx)
    require_dispute(
        merkle_check(
            state_root,
            commit.fee_index,
            dispute.fee_account_post,
            dispute.fee_witness_post
        )
    )
    state_root = apply_fee(dispute.fee_account, fee_sum)
```

### PR DESCRIPTION
**Pull request by ChihChengLiang**
Fix #206

### CODE REVIEW

File: contracts/Create2Transfer.sol [hunk 1]
```diff
@@ -86,11 +86,12 @@ contract Create2Transfer {
         bytes32 stateRoot,
         bytes memory txs,
         Types.StateMerkleProof[] memory proofs,
-        uint256 tokenType,
         uint256 feeReceiver
     ) public pure returns (bytes32, Types.Result result) {
         uint256 length = txs.create2Transfer_size();
         uint256 fees = 0;
+        // tokenType should be the same for all states in this commit
+        uint256 tokenType = proofs[0].state.tokenType;
```
Comment (ChihChengLiang) [inline contracts/Create2Transfer.sol:94]: We are taking the first tokenType provided by the disputer. In the first validateSender the tokenType will trivially match. This tokenType can be trusted if it passed the merkle inclusion check in the first validateSender. 

Then in the later processTx, this tokenType can use to verify if all of the states of sernders, receivers, and fee receivers, have the same tokenType.

Comment (vaibhavchellani) [reply contracts/Create2Transfer.sol:94]: Yeah that makes a lot of sense

**Review by vaibhavchellani — APPROVED** [review]
LGTM

**Comment by vaibhavchellani** [comment]
Should we remove it from other tx types too?

**Comment by ChihChengLiang** [comment]
@vaibhavchellani in this PR we removed tokenType commitment from "Transfer" and "Create2Transfer". We have to keep the MassMigration's tokenType commitment because the withdraw manager has to know which token to withdraw. We can remove that tokenType commitment, but will come at a price of adding a merkle proof check for state.

**Comment by vaibhavchellani** [comment]
Gotcha, that makes sense, I guess we can keep mm's as they are

**Pull request merged**